pkg/util/log/line_buffer.go

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/grafana/loki/blob/7c78d7ea44afb420847255f9f5a4f677ad0f47bf/pkg/util/log/line_buffer.go
+// Provenance-includes-copyright: Grafana Labs
+package log
+
+import (
+	"bytes"
+	"io"
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+// LineBufferedLogger buffers log lines to be flushed periodically. Without a line buffer, Log() will call the write
+// syscall for every log line which is expensive if logging thousands of lines per second.
+type LineBufferedLogger struct {
+	buf     *threadsafeBuffer
+	entries atomic.Uint32
+	cap     uint32
+	w       io.Writer
+
+	onFlush func(entries uint32)
+}
+
+// Size returns the number of entries in the buffer.
+func (l *LineBufferedLogger) Size() uint32 {
+	return l.entries.Load()
+}
+
+// Write writes the given bytes to the line buffer, and increments the entries counter.
+// If the buffer is full (entries == cap), it will be flushed, and the entries counter reset.
+func (l *LineBufferedLogger) Write(p []byte) (n int, err error) {
+	// when we've filled the buffer, flush it
+	if l.Size() >= l.cap {
+		// Flush resets the size to 0
+		if err := l.Flush(); err != nil {
+			l.buf.Reset()
+			return 0, err
+		}
+	}
+
+	l.entries.Inc()
+
+	return l.buf.Write(p)
+}
+
+// Flush forces the buffer to be written to the underlying writer.
+func (l *LineBufferedLogger) Flush() error {
+	sz := l.Size()
+	if sz <= 0 {
+		return nil
+	}
+
+	// reset the counter
+	l.entries.Store(0)
```

Comment:
I'd say we should call `sz := l.entries.Swap(0)` here and only proceed is `sz > 0` to avoid calling `l.buf.WriteTo` and `onFlush` twice.  While second `l.buf.WriteTo` is harmless (buffer is empty), second `l.onFlush(sz)` is incorrect.

Reply:
Where is `onFlush` being called a second time?

Reply:
Sorry for confusion. I was thinking about two goroutines that enter `Flush` method concurrently. If they both will detect positive `sz` and don't return early, then `l.onFlush(sz)` will be called twice with same value -- but only one goroutine can possibly write data in `l.buf.WriteTo(l.w)` call.

Alternative fix is to use value returned from `l.buf.WriteTo(l.w)` and report that to `l.onFlush(...)` (if it's non-zero).

Reply:
`l.buf.WriteTo(l.w)` returns the size in bytes, and `onFlush` expects the number of flushed entries instead.

Reply:
Reading this, I'm also wondering if it isn't cleaner/more robust wrt. concurrency to do what Peter suggests?